server/src/main/java/org/opensearch/index/query/SourceFieldMatchQuery.java

```diff
@@ -0,0 +1,161 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.query;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.memory.MemoryIndex;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TwoPhaseIterator;
+import org.apache.lucene.search.Weight;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.mapper.SourceValueFetcher;
+import org.opensearch.search.lookup.LeafSearchLookup;
+import org.opensearch.search.lookup.SearchLookup;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A query that matches against each document from the parent query by filtering using the source field values.
+ * Useful to query against field type which doesn't store positional data and field is not stored/computed dynamically.
+ */
+public class SourceFieldMatchQuery extends Query {
+    final private Query delegateQuery;
+    final private Query filter;
+    final private SearchLookup lookup;
+    final private MappedFieldType fieldType;
+    final private SourceValueFetcher valueFetcher;
+    final private QueryShardContext context;
+
+    /**
+     * Constructs a SourceFieldMatchQuery.
+     *
+     * @param delegateQuery The parent query to use to find matches.
+     * @param filter The query used to filter further by running against field value fetched using _source field.
+     * @param fieldType The mapped field type.
+     * @param context The QueryShardContext to get lookup and valueFetcher
+     */
+    public SourceFieldMatchQuery(Query delegateQuery, Query filter, MappedFieldType fieldType, QueryShardContext context) {
+        this.delegateQuery = delegateQuery;
+        this.filter = filter;
+        this.fieldType = fieldType;
+        this.context = context;
+        this.lookup = context.lookup();
+        if (!context.documentMapper("").sourceMapper().enabled()) {
+            throw new IllegalArgumentException(
+                "SourceFieldMatchQuery error: unable to fetch fields from _source field: _source is disabled in the mappings "
+                    + "for index ["
+                    + context.index().getName()
+                    + "]"
+            );
+        }
+        this.valueFetcher = (SourceValueFetcher) fieldType.valueFetcher(context, lookup, null);
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        delegateQuery.visit(visitor);
+    }
+
+    @Override
+    public Query rewrite(IndexSearcher indexSearcher) throws IOException {
```

Comment:
Don't we have to call `super.rewrite` to trigger the `Query` class's `rewrite` method?